### PR TITLE
[Agent] Inject index with sane defaults to configuration

### PR DESCRIPTION
**Pull request by michalpristas**
Migrated from: #15852
Build on top of: #15809

Added one extra rule for injecting index into inputs based on provided type and detected namespace/dataset type.

Other option was to use decorator but at the time of decorating there are already some pieces of config missing. 

Draft until final configuration format is settled and mentioned PR is in, but READY for review.

It overwrites index specified in input. 

Fixes: #15690
Fixes: #15691

### CODE REVIEW

File: x-pack/agent/pkg/agent/program/testdata/single_config-filebeat.yml [hunk 1]
```diff
@@ -4,10 +4,13 @@ filebeat:
...
+    index: filebeat-generic-default
 output:
   elasticsearch:
     hosts:
       - 127.0.0.1:9200
       - 127.0.0.1:9300
     username: elastic
     password: changeme
+    api_key:	TiNAGG4BaaMdaH1tRfuU:KnR6yE41RrSowb0kQ0HWoA
+    ca_sha256:	7HIpactkIAq2Y49orFOOQKurWxmmSFZhBCoQYcRhJ3Y=
```
Comment (ph) [inline x-pack/agent/pkg/agent/program/testdata/single_config-filebeat.yml]: Space looks weird here?

File: x-pack/agent/pkg/agent/program/testdata/single_config-filebeat.yml [hunk 2]
```diff
@@ -4,10 +4,13 @@ filebeat:
     paths:
       - /var/log/hello1.log
       - /var/log/hello2.log
+    index: filebeat-generic-default
```
Comment (ph) [inline x-pack/agent/pkg/agent/program/testdata/single_config-filebeat.yml]: This is incorrect, it should be: 
```
logs-{dataset}-{namespace}
```

@ruflin When no dataset and namespace are defined what are the default values for them? The https://github.com/elastic/kibana/pull/56132 doesn't mention it?

File: x-pack/agent/pkg/agent/program/testdata/single_config-metricbeat.yml [hunk 3]
```diff
@@ -1,11 +1,15 @@
 metricbeat:
   modules:
   - module: docker
-    setting: one
+    metricsets: [status]
+    index: metricbeat-docker.status-default
   - module: apache
-    setting: two
+    metricsets: [info]
+    index: metricbeat-generic-testing
```
Comment (ph) [inline x-pack/agent/pkg/agent/program/testdata/single_config-metricbeat.yml]: should be metrics-{dataset}-{namespace}

File: x-pack/agent/pkg/agent/program/testdata/single_config.yml [hunk 4]
```diff
@@ -1,19 +1,49 @@
...
-      - /var/log/hello2.log
-management:
-  host: "localhost"
-config:
-  reload: 123
+name:	Production Website DB Servers
+fleet:
+    kibana_url:		https://kibana.mydomain.com:5601
+    ca_hash:		7HIpactkIAq2Y49orFOOQKurWxmmSFZhBCoQYcRhJ3Y=
+    checkin_interval:	5m
```
Comment (ph) [inline x-pack/agent/pkg/agent/program/testdata/single_config.yml]: Something weird with the same in a few place, I think YAML will strip it?

File: x-pack/agent/pkg/agent/program/testdata/single_config-filebeat.yml (hunk 2 above)
Comment (ruflin) [reply x-pack/agent/pkg/agent/program/testdata/single_config-filebeat.yml]: https://github.com/elastic/kibana/pull/56132/files#diff-abbe71b916f44ea720e680dd923aba1cR101

Comment (michalpristas) [reply x-pack/agent/pkg/agent/program/testdata/single_config-filebeat.yml]: this issue mentions defaults #15691
for NS it is `default` for dataset `generic` so this follows `logs-dataset-namespace` correctly

File: x-pack/agent/pkg/agent/program/testdata/single_config-filebeat.yml (hunk 1 above)
Comment (michalpristas) [reply x-pack/agent/pkg/agent/program/testdata/single_config-filebeat.yml]: it does indeed

File: x-pack/agent/pkg/agent/program/testdata/single_config-metricbeat.yml (hunk 3 above)
Comment (michalpristas) [reply x-pack/agent/pkg/agent/program/testdata/single_config-metricbeat.yml]: namespace==testing in this case so i think it is correct.

File: x-pack/agent/pkg/agent/program/testdata/single_config-filebeat.yml (hunk 2 above)
Comment (ruflin) [reply x-pack/agent/pkg/agent/program/testdata/single_config-filebeat.yml]: The above should be `logs-generic-default`.

Comment (ph) [reply x-pack/agent/pkg/agent/program/testdata/single_config-filebeat.yml]: I believe I've read it too fast. thanks

File: x-pack/agent/_meta/agent.docker.yml [hunk 5]
```diff
@@ -8,12 +8,21 @@ outputs:
...
+    inputs:
+      - type: logs
+        processors:
+        streams:
+          - dataset: sample.acccess
+            paths: /var/log/sample/access.log
+          - enabled: false
+            dataset: sample.error
+            paths: /var/log/sample/error.log
+      - type: metrics/system
```
Comment (ph) [inline x-pack/agent/_meta/agent.docker.yml]: it should be `system/metrics` as defined in https://github.com/elastic/beats/blob/feature-ingest/x-pack/agent/docs/agent_configuration_example.yml#L56

File: x-pack/agent/pkg/agent/program/testdata/single_config.yml (hunk 4 above)
Comment (michalpristas) [reply x-pack/agent/pkg/agent/program/testdata/single_config.yml]: this is weird i have a single space here

File: x-pack/agent/pkg/agent/transpiler/rules.go [hunk 6]
```diff
@@ -143,6 +159,214 @@ func (r *RuleList) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+// namespace: is collected from streams[n].namespace. If not found used 'default'.
+// dataset-type: is collected from streams[n].dataset.type. If not found used 'generic'.
+type InjectIndexRule struct {
+	Type string
+}
+
+// Apply injects index into input.
+func (r *InjectIndexRule) Apply(ast *AST) error {
+	const defaultNamespace = "default"
+	const defaultDataset = "generic"
```
Comment (ph) [inline x-pack/agent/pkg/agent/transpiler/rules.go]: ++ to have theses constants in the function.

**Comment by michalpristas** [comment]
going to test with kibana, either with working UI or manual API calls

**Comment by michalpristas** [comment]
@ruflin  @ph tested with kibana ok, i will create a separate PR targeting master

**Comment by michalpristas** [comment]
created new PR against master here: https://github.com/elastic/beats/pull/16903
closing this one

**Pull request closed without merging**